src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs

```diff
@@ -491,6 +491,62 @@ internal SynthesizedAttributeData SynthesizeDynamicAttribute(TypeSymbol type, in
             }
         }
 
+        internal SynthesizedAttributeData SynthesizeTupleNamesAttribute(TypeSymbol type)
+        {
+            Debug.Assert((object)type != null);
+            Debug.Assert(type.ContainsTuple());
+
+            var stringType = GetSpecialType(SpecialType.System_String);
+            Debug.Assert((object)stringType != null);
+            var names = TupleNamesEncoder.Encode(type, stringType);
+            var stringArray = ArrayTypeSymbol.CreateSZArray(stringType.ContainingAssembly, stringType, customModifiers: ImmutableArray<CustomModifier>.Empty);
+            var args = ImmutableArray.Create(new TypedConstant(stringArray, names));
+            return TrySynthesizeAttribute(WellKnownMember.System_Runtime_CompilerServices_TupleElementNamesAttribute__ctorTransformNames, args);
+        }
+
+        private static class TupleNamesEncoder
+        {
+            public static ImmutableArray<TypedConstant> Encode(TypeSymbol type, TypeSymbol stringType)
+            {
+                var namesBuilder = ArrayBuilder<string>.GetInstance();
+                type.VisitType((t, builder, _ignore) => AddName(t, builder), namesBuilder);
+                Debug.Assert(namesBuilder.Any());
```

Comment:
When generating `[DynamicAttribute(new bool[] { ... })]` do we drop trailing 'false' values? (I believe metadata decoding allows the array to be shorter than the number of elements in the type and treats the missing values as 'false'.) Should we do the same here for trailing 'null' values on encoding and decoding?